pkgpanda/build/__init__.py

```diff
@@ -100,6 +100,111 @@ def get_src_fetcher(src_info, cache_dir, working_directory):
         raise BuildError("Validation error when fetching sources for package: {}".format(ex))
 
 
+class TreeInfo:
+
+    ALLOWED_TREEINFO_KEYS = {'exclude', 'variants', 'core_package_list', 'bootstrap_package_list'}
+
+    def __init__(self, treeinfo_dict):
+        if treeinfo_dict.keys() > self.ALLOWED_TREEINFO_KEYS:
+            raise BuildError(
+                "treeinfo can only include the keys {}. Found {}".format(
+                    self.ALLOWED_TREEINFO_KEYS, treeinfo_dict.keys()))
+
+        self.excludes = set(self._get_package_list(treeinfo_dict, 'exclude'))
+        self.core_package_list = set(self._get_package_list(treeinfo_dict, 'core_package_list', self.excludes))
+        self.bootstrap_package_list = set(self._get_package_list(
+            treeinfo_dict,
+            'bootstrap_package_list',
+            self.excludes))
+
+        # List of mandatory package variants to include in the buildinfo.
+        self.variants = treeinfo_dict.get('variants', dict())
+        if not isinstance(self.variants, dict):
+            raise BuildError("treeinfo variants must be a dictionary of package name to variant name")
+
+    @staticmethod
+    def _get_package_list(treeinfo_dict, key, excludes=None):
+        """Return a list of package name strings from treeinfo_dict by key.
+
+        If key isn't present in treeinfo_dict, an empty list is returned.
+
+        """
+        excludes = excludes or list()
+        package_list = treeinfo_dict.get(key, list())
+
+        # Validate package list.
+        if not isinstance(package_list, list):
+            raise BuildError("{} must be either null (meaning don't use) or a list of package names.".format(key))
+        for package_name in package_list:
+            if not isinstance(package_name, str):
+                raise BuildError("{} must be a list of strings. Found a {} with the value: {}".format(
+                    key, type(package_name), package_name))
+
+            if package_name in excludes:
+                raise BuildError("Package found in both exclude and {}: {}".format(key, package_name))
+
+        return package_list
+
+
+class PackageSet:
+
+    def __init__(self, variant, treeinfo, package_store):
+        self.variant = variant
+
+        all_unexcluded_packages = package_store.packages_by_name.keys() - treeinfo.excludes
+
+        self.all_packages = self.package_tuples_with_dependencies(
+            treeinfo.core_package_list or all_unexcluded_packages,
+            treeinfo,
+            package_store
+        )
+        self.validate_package_tuples(self.all_packages, treeinfo, package_store)
+
+        self.bootstrap_packages = self.package_tuples_with_dependencies(
```

Comment:
I'll go with that, it saves some work if the treeinfo doesn't have any bootstrap packages.